Allow error rename and disallow error rename for all AWS protocols (#1554)
awslabs/smithy: docs/source-2.0/aws/protocols/aws-json-1_1-protocol.rst

.. |protocol test link| replace:: https://github.com/awslabs/smithy/tree/main/smithy-aws-protocol-tests/model/awsJson1_1
.. include:: aws-json.rst.template

> --------------------

Reviewer: I think we should prohibit renames for each AWS protocol, not just aws-json-1.1. That would probably mean you'd want to break this content out into an include thing with a .txt file to not have to copy/paste so much in the docs of each protocol. That or make a section at the top level of the protocol docs to document this and link to it from each protocol doc (e.g., here https://smithy.io/2.0/aws/protocols/index.html)

Author: Makes sense. I have simplified the documentation for now and not go into the error shape ID/name details because each protocol behaves differently.

Error shape renaming
--------------------

> By default, Smithy permits renaming shapes to disambiguate shape ID conflicts in

Reviewer: I think this paragraph and the next paragraph should be combined (we don't need this many one sentence paragraphs).

the :ref:`service closure <service-closure>` via the ``rename`` property.

However, services using the ``awsJson1_1`` protocol are not
allowed to rename error shapes (shapes with :ref:`error trait <error-trait>` applied).

> According to ``Operation error serialization``,

Reviewer: Is this supposed to be a link? Right now it's just `<code>` text.

> client-side implementation relies on the response body field ``code`` or ``__type`` to resolve the error type.

Reviewer: ```suggestion
client-side implementations rely on the response body field ``code`` or ``__type`` to resolve the error type.
```

> Server-side implementation of ``awsJson1_1`` protocol will only send the ``shape name`` for the response body field.

Reviewer: ```suggestion
Server-side implementations of ``awsJson1_1`` protocol will only send the ``shape name`` for the response body field.
```

When there are conflicting shape IDs ``smithy.service#ServiceError`` and ``smithy.other#ServiceError``,

> server-side will only send the shape name ``ServiceError``. Client-side implementation will not be able to resolve

Reviewer: ```suggestion
the server will only send the shape name ``ServiceError``. Clients will not be able to resolve
```

> the correct error type without the ``namespace``.

Reviewer: ```suggestion
the correct error type without the namespace.
```

> Server-side implementation of ``awsJson1_1`` protocol doesn't handle or send renamed shape names.

Reviewer: ```suggestion
Server-side implementations of ``awsJson1_1`` don't serialize renamed shape names.
```

As a result, renaming will not resolve the conflicting shape IDs issue, and hence it is not permitted.

> The resolution should be to avoid having conflicting error shape IDs.

Reviewer: Let's remove this sentence
